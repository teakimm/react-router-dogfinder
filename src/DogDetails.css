.DogDetails {
  display: flex;
  justify-content: center;
  margin-top: 2em;
}

ul {
  margin-left: 2em;
  text-align: left;
}