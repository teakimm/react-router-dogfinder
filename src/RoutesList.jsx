import React from "react";
import { Navigate, Route, Routes } from "react-router-dom";
import DogList from "./DogList";
import DogDetails from "./DogDetails";
import { useParams } from "react-router-dom";

/** Routes to render components
 *
 * state:
 * none
 *
 * props:
 * -dogs: [{name, age, ...}]
 *
 * App -> RoutesList -> DogDetails, DogList
 */
function RoutesList({ dogs }) {


    return (
        <Routes>
            <Route path="*" element={<Navigate to="/dogs" />} />
            <Route path="dogs/:name" element={<DogDetails dogs={dogs} />} />
            <Route path="/dogs" element={<DogList dogs={dogs} />} />
        </Routes>
    );
}

export default RoutesList;