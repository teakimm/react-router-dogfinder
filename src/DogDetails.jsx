import './App.css';
import { BrowserRouter, Link, Navigate, Route } from 'react-router-dom';
import RoutesList from './RoutesList';
import { useState } from 'react';
import { v4 as uuid } from "uuid";
import { useParams } from 'react-router-dom';
import "./DogDetails.css";
import NotFound from './NotFound';

/** Renders card for a dog
 *
 * state:
 * none
 *
 * props:
 * -dogs [{name, age, ...}]
 *
 * App -> RouteList -> DogDetails
 */
function DogDetails({ dogs }) {
    const { name } = useParams();
    const dog = dogs.find(dog => dog.name.toLowerCase() === name.toLowerCase());

    if (!dog) {
        return <NotFound />;
    }

    return (
        <div className="DogDetails">
            <div className="card DogDetails-card" style={{ width: '25em' }}>
                <img className="card-img-top" src={`/public/${dog.src}.jpg`} alt={dog.src} />
                <div className="card-body">
                    <h5 className="card-title">{dog.name}</h5>
                    <h6>Age: {dog.age}</h6>
                    <ul>
                        {dog.facts.map((fact, idx) => <li key={idx}>{fact}</li>)}
                    </ul>
                    <Link className="btn btn-primary" to={`/dogs`}>Go Back</Link>
                </div>
            </div>
        </div>
    );
};
export default DogDetails;