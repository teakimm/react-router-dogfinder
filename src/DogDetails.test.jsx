import { describe, it, expect, vi } from "vitest";

import React from "react";
import { render, fireEvent } from "@testing-library/react";
import DogDetails from "./DogDetails";
import "../db.json";
import { MemoryRouter } from "react-router-dom";
import RoutesList from "./RoutesList";

const DOGS = [
    {
        "name": "Whiskey",
        "age": 5,
        "src": "whiskey",
        "facts": [
            "Whiskey loves eating popcorn.",
            "Whiskey is a terrible guard dog.",
            "Whiskey wants to cuddle with you!"
        ]
    },
    {
        "name": "Duke",
        "age": 3,
        "src": "duke",
        "facts": [
            "Duke believes that ball is life.",
            "Duke likes snow.",
            "Duke enjoys pawing other dogs."
        ]
    },
    {
        "name": "Perry",
        "age": 4,
        "src": "perry",
        "facts": [
            "Perry loves all humans.",
            "Perry demolishes all snacks.",
            "Perry hates the rain."
        ]
    }
];

describe("DogDetails component", function () {
    it("renders without crashing", function () {
        const { container, debug } = render(
            <MemoryRouter initialEntries={["/dogs/Perry"]}>
                <RoutesList dogs={DOGS} />
            </MemoryRouter>
        );
    });

    it("matches snapshot", function () {
        const { container, debug } = render(
            <MemoryRouter initialEntries={["/dogs/Perry"]}>
                <RoutesList dogs={DOGS} />
            </MemoryRouter>
        );
        expect(container).toMatchSnapshot();
    });

    it("renders correct dog information", function () {
        const result = render(
            <MemoryRouter initialEntries={["/dogs/Perry"]}>
                <RoutesList dogs={DOGS} />
            </MemoryRouter>
        );

        expect(result.queryByText("Perry loves all humans.")).toBeInTheDocument();
        expect(result.queryByText("Duke likes snow.")).not.toBeInTheDocument();
    });

    it("displays 404 on non-existent dog", function () {
        const result = render(
            <MemoryRouter initialEntries={["/dogs/David"]}>
                <RoutesList dogs={DOGS} />
            </MemoryRouter>
        );

        expect(result.queryByText(
            "We couldn't find the dog you were looking for.")).toBeInTheDocument();
    });
});