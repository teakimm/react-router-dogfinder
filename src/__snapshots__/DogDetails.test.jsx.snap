// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DogDetails component > matches snapshot 1`] = `
<div>
  <div
    class="DogDetails"
  >
    <div
      class="card DogDetails-card"
      style="width: 25em;"
    >
      <img
        alt="perry"
        class="card-img-top"
        src="/public/perry.jpg"
      />
      <div
        class="card-body"
      >
        <h5
          class="card-title"
        >
          Perry
        </h5>
        <h6>
          Age: 
          4
        </h6>
        <ul>
          <li>
            Perry loves all humans.
          </li>
          <li>
            Perry demolishes all snacks.
          </li>
          <li>
            Perry hates the rain.
          </li>
        </ul>
        <a
          class="btn btn-primary"
          href="/dogs"
        >
          Go Back
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`DogDetails component > renders without crashing 1`] = `
<div>
  <div
    class="DogDetails"
  >
    <div
      class="card DogDetails-card"
      style="width: 25em;"
    >
      <img
        alt="perry"
        class="card-img-top"
        src="/public/perry.jpg"
      />
      <div
        class="card-body"
      >
        <h5
          class="card-title"
        >
          Perry
        </h5>
        <h6>
          Age: 
          4
        </h6>
        <ul>
          <li>
            Perry loves all humans.
          </li>
          <li>
            Perry demolishes all snacks.
          </li>
          <li>
            Perry hates the rain.
          </li>
        </ul>
        <a
          class="btn btn-primary"
          href="/dogs"
        >
          Go Back
        </a>
      </div>
    </div>
  </div>
</div>
`;
